File: packages/javascript/lib/node/os/getSid.ts
import { homedir, platform, userInfo } from 'node:os';
import { runCommand } from './runCommand.js';

const PROFILE_LIST_KEY = 'HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\ProfileList';

// `REG QUERY` writes the name, the type and the value of a value with runs of
// spaces between them. A profile path holds single spaces of its own, so the run
// is what separates the columns, not a fixed count.
const PROFILE_PATH_LINE = /^ProfileImagePath\s{2,}REG_\w+\s{2,}(.+)$/;

export async function getSid(): Promise<string> {
	const platformName = platform();

	if (platformName === 'win32') {
		return getWindowsSid();
	}

	if (platformName === 'darwin') {
		return getDarwinSid();
	}

	throw new Error('Not supported on this operating system.');
}

async function getWindowsSid(): Promise<string> {
	// Every profile on the machine is listed under a key named after its SID, and
	// carries the folder that profile owns. The one owning this user's home
	// directory is this user's SID.
	const profileList = await runCommand(`REG QUERY "${PROFILE_LIST_KEY}" /s`);
	const home = homedir().toLowerCase();

	let currentSid = '';

	for (const line of (profileList || '').split(/\r?\n/)) {
		const trimmed = line.trim();

		if (trimmed.startsWith('HKEY_LOCAL_MACHINE\\')) {
			currentSid = trimmed.split('\\').pop() || '';
			continue;
		}

		const profilePath = trimmed.match(PROFILE_PATH_LINE);

		// Windows paths are compared without case, so a profile recorded as
		// `C:\Users\Sam` still matches a home directory read as `C:\users\sam`.
		if (currentSid && profilePath && profilePath[1].trim().toLowerCase() === home) {
			return currentSid;
		}
	}

	throw new Error('Failed to get the SID of the current user');
}

async function getDarwinSid(): Promise<string> {
	// Quoted because the name comes from the system rather than from this code, and
	// it is written into a shell command. macOS does not allow a single quote in a
	// user name, so there is nothing for the quotes to fail to contain.
	const sid = await runCommand(`dsmemberutil getsid -U '${userInfo().username}'`);

	if (!sid) {
		throw new Error('Failed to get the SID of the current user');
	}

	return sid.replace(/\r?\n/g, '');
}

File: packages/python/qsu/os/getSid.py
import getpass
import os
import re
import sys

from .runCommand import runCommand

_PROFILE_LIST_KEY = r'SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList'


def getSid() -> str:
	if sys.platform == 'win32':
		return _windowsSid()

	if sys.platform == 'darwin':
		return _darwinSid()

	raise RuntimeError('Not supported on this operating system.')


def _windowsSid() -> str:
	if sys.platform != 'win32':
		raise RuntimeError('Not supported on this operating system.')

	# Every profile on the machine is listed under a key named after its SID, and
	# carries the folder that profile owns. The one owning this user's home directory
	# is this user's SID. The registry is read directly, so nothing is spawned and
	# there is no command output to parse.
	import winreg

	home = os.path.expanduser('~').lower()

	try:
		with winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, _PROFILE_LIST_KEY) as profiles:
			for index in range(winreg.QueryInfoKey(profiles)[0]):
				sid = winreg.EnumKey(profiles, index)

				try:
					with winreg.OpenKey(profiles, sid) as profile:
						path = winreg.QueryValueEx(profile, 'ProfileImagePath')[0]
				except OSError:
					continue

				# Windows paths are compared without case, so a profile recorded as
				# `C:\Users\Sam` still matches a home directory read as `C:\users\sam`.
				if isinstance(path, str) and path.strip().lower() == home:
					return sid
	except OSError as error:
		raise RuntimeError(str(error))

	raise RuntimeError('Failed to get the SID of the current user')


def _darwinSid() -> str:
	# Quoted because the name comes from the system rather than from this code, and
	# it is written into a shell command. macOS does not allow a single quote in a
	# user name, so there is nothing for the quotes to fail to contain.
	sid = runCommand(f"dsmemberutil getsid -U '{getpass.getuser()}'")

	if not sid:
		raise RuntimeError('Failed to get the SID of the current user')

	return re.sub(r'\r?\n', '', sid)
